# This notebook shows how to automatically load today's games, and then predict/tweet said forecast

In [1]:
#Step 1. Load in the data and dependencies. 
%matplotlib inline
import numpy as np
import NNBA
import prediction_stuff
import matplotlib.pyplot as plt
np.random.seed(11)  #For reproducibility. And because Brook Lopez is the best. 
#note there is continuation, so if you run it twice without resetting the kernel the restuls will be different. 

/Users/noahkasmanoff/anaconda/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
Using TensorFlow backend.


In [2]:

#Create a model for predicting the spread outcome of a game.

model,scaler = NNBA.make_network('NBADATA.csv',sklearn=False,
                                          keras=True,normalize=True,
                                          spread=True)




keras NN goes here
Train on 3820 samples, validate on 955 samples
Epoch 1/20
3820/3820 [==============================] - 1s 186us/step - loss: 0.8511 - acc: 0.4890 - val_loss: 0.7860 - val_acc: 0.4838
Epoch 2/20
3820/3820 [==============================] - 0s 41us/step - loss: 0.7247 - acc: 0.6270 - val_loss: 0.6986 - val_acc: 0.6848
Epoch 3/20
3820/3820 [==============================] - 0s 40us/step - loss: 0.5597 - acc: 0.7644 - val_loss: 0.5115 - val_acc: 0.8084
Epoch 4/20
3820/3820 [==============================] - 0s 40us/step - loss: 0.4313 - acc: 0.8259 - val_loss: 0.4552 - val_acc: 0.8188
Epoch 5/20
3820/3820 [==============================] - 0s 40us/step - loss: 0.3818 - acc: 0.8555 - val_loss: 0.4234 - val_acc: 0.8314
Epoch 6/20
3820/3820 [==============================] - 0s 39us/step - loss: 0.3543 - acc: 0.8623 - val_loss: 0.4726 - val_acc: 0.8031
Epoch 7/20
3820/3820 [==============================] - 0s 39us/step - loss: 0.3436 - acc: 0.8678 - val_loss: 0.3961 - val_

This printed result means that when it scanned over NBADATA.csv, a file created using nba_py (documented in NNBA.py), after training over 3820 games and testing over 955 games, the nbapredictor neural network could correctly classify whether or not an nba game hit the over or under 86% of the time.

Note this is based on the box scores of existing games! The challenge now is to find a way to predict what the box score of a future NBA game will turn out to be. In the cells below, we use the season splits of the Cleveland Cavaliers and Golden State Warriors to predict the spread winner of each game of the 2018 NBA Finals. 

In the end, this model ends up successfully --or the highest % belonging to the true winner--  4 out of the 4 times! I'd still consult other sources for game 3 though. Granted a small sample size, this model shows promise when extended throughout the entire NBA playoffs, and we look forward to seeing what comes of it start the 2018-19 NBA season. 

Of course there are many flaws with this system as the box score of a game will almost never be the exact regular season splits of a team. But this serves as a good starting point in learning more about NBA game prediction and modelling. 


And for other types of outcomes, the over under and ML are still experimental but follow the same procedure. Neither are as coherently programmed as this one so this is the launching point of our project. 

In [ ]:
CLE = prediction_stuff.get_splits('CLE')
GSW = prediction_stuff.get_splits('GSW')
spread1 = 13
spread2 = 12.5
spread3 = -3.5
spread4 = -4.5
spreads = [spread1,spread2,spread3,spread4]

CLEGSW1 = prediction_stuff.ML_game_maker(CLE,GSW,spread1,scaler)
CLEGSW2 = prediction_stuff.ML_game_maker(CLE,GSW,spread2,scaler)
CLEGSW3 = prediction_stuff.ML_game_maker(GSW,CLE,spread3,scaler)
CLEGSW4 = prediction_stuff.ML_game_maker(GSW,CLE,spread4,scaler)

the_finals = [CLEGSW1,CLEGSW2,CLEGSW3,CLEGSW4]

for i,game in enumerate(the_finals):
    prediction =  nbapredictorML.predict(game)
    plt.figure()
    # Data to plot
    labels = 'Cavs', 'Warriors'
    sizes = prediction[0]
    colors = ['lightcoral', 'gold', 'green']
    explode = (0.1, 0.1)  # explode each slice
 
    # Plot
    plt.pie(sizes, explode=explode, labels=labels, colors=colors,
        autopct='%1.1f%%', shadow=True, startangle=140)
 
    plt.axis('equal')
    plt.title("Who Wins Game "+ str(i+1) + "?")

In [ ]:
from datetime import datetime

In [ ]:
now = datetime.now()

In [ ]:
now.year

In [ ]:
matchups = prediction_stuff.get_todays_games(now.day,now.month,now.year)

In [3]:
matchups = prediction_stuff.get_todays_games(30,3,2017)
print(matchups)
teamsplits = []
for game in matchups:
    for team in game:
        teamsplit = prediction_stuff.get_splits(team)  #this is one of the arrays, now pair it up like was done
        teamsplits.append(teamsplit)

home_teams = teamsplits[1::2]
away_teams = teamsplits[0::2]


#Spread will be extracted for that game...
todays_matchups = []
for i in range(len(home_teams)):
    todays_matchups.append(prediction_stuff.spread_game_maker(away_teams[i],home_teams[i],7,scaler))

for game in todays_matchups:
    print(model.predict(game))



Old Teams: ['BKN', 'DET', 'CLE', 'CHI', 'LAL', 'MIN', 'LAC', 'PHX', 'HOU', 'POR']
Fixed(?) Teams:  ['BRK', 'DET', 'CLE', 'CHI', 'LAL', 'MIN', 'LAC', 'PHX', 'HOU', 'POR']
[['BRK', 'DET'], ['CLE', 'CHI'], ['LAL', 'MIN'], ['LAC', 'PHX'], ['HOU', 'POR']]
[[ 0.73748505  0.21747054  0.04504435]]
[[ 0.95844471  0.03281399  0.00874139]]
[[ 0.98868352  0.00658248  0.00473394]]
[[ 0.99311757  0.00391528  0.00296713]]
[[ 0.92713326  0.05568774  0.01717897]]


In [ ]:
teams = ['BKN', 'DET', 'CLE', 'CHI', 'LAL', 'MIN', 'LAC', 'PHX', 'HOU', 'POR']

In [ ]:
new_teams = []
for team in teams:
    if team == 'BKN':
        new_teams.append('BRK')
    else:
        new_teams.append(team)

In [ ]:
new_teams